src/util/db/syncing.ts

```diff
@@ -0,0 +1,188 @@
+
+      // Query for documents
+      const keys = Object.keys(changeSet).map(path => `${repoId}:${path}`)
+      const fetchResponse = await appState.dataStore.fetch({ keys })
+
+      // Prepare updated documents
+      const updatedStoreFileDocs = fetchResponse.rows.map((row, i) => {
+        const path = row.key.split(':')[1]
+        const box = changeSet[path]
+        const newVersion = latestCheckpoint.version + i + 1
+
+        if (row.error != null && row.error !== 'not_found') {
+          throw new Error(row.error)
+        }
+
+        const existingStoreFileDoc =
+          'doc' in row && row.doc != null
+            ? asMaybe(asStoreFileDocument)(row.doc)
+            : undefined
+        const templateStoreFileDoc = existingStoreFileDoc ?? {
+          _id: row.key,
+          box,
+          timestamp: Date.now(),
+          versions: []
+        }
+
+        return {
+          ...templateStoreFileDoc,
+          box,
+          timestamp: Date.now(),
+          versions: [newVersion, ...templateStoreFileDoc.versions].slice(
+            0,
+            MAX_DOCUMENT_VERSION_COUNT
+          )
+        }
+      })
+
+      // Bulk update
+      const updateResponse = await appState.dataStore.bulk({
+        docs: updatedStoreFileDocs
+      })
+
+      // Check for any errors from update
+      checkDbResponseForErrors(updateResponse)
+    },
+    err => err.message === 'conflict'
+  )
+
+export interface RepoUpdates {
+  checkpoints: CheckpointArray
+  changeSet: ChangeSetV2
+}
+
+export const readUpdates = (appState: AppState) => async (
+  repoId: string,
+  checkpoints: CheckpointArray
+): Promise<RepoUpdates> => {
+  const defaultCheckpoint = { version: 0, sum: 0 }
+  const currentCheckpoint = checkpoints[0] ?? defaultCheckpoint
+
+  // Query view for file documents with version > currentCheckpoint.version
+  const fileDocumentsResponse = await appState.dataStore.partitionedView(
+    repoId,
+    'versioning',
+    'version',
+    {
+      include_docs: true,
+      reduce: false,
+      // Add 1 because range is exclusive
+      start_key: currentCheckpoint.version + 1
+    }
+  )
+
+  const storeFileDocs = fileDocumentsResponse.rows.map(row => {
+    return trial(
+      () => asStoreFileDocument(row.doc),
+      err => {
+        throw new Effect(err, `Expected StoreFile document for ${row.id}`)
+      }
+    )
+  })
+
+  // Query view for checkpoint sum where version ≤ currentCheckpoint.version
+  const { sum: wipSum } = await getCheckpointAt(appState)(
+    repoId,
+    currentCheckpoint.version
+  )
+
+  // The expected difference between client and server sums
+  // caused by overwritten documents.
+  const overwriteDiff = storeFileDocs.reduce((sum, storeFile) => {
+    const historicalVersions = storeFile.versions
+      .slice(1)
+      .filter(version => version <= currentCheckpoint.version)
+
+    return sum + Math.max(0, ...historicalVersions)
```

Comment:
If the versions are in order, we should just be able to grab the first one, instead of using `Math.max`?

Reply:
Is this resolved? I believe it is.